apps/shipments/permissions.py

```diff
@@ -33,22 +31,51 @@ class IsOwnerOrShared(permissions.IsAuthenticated):
     Custom permission to allow users with a unique link to access a shipment anonymously
     """
     def has_permission(self, request, view):
-        return (super(IsOwnerOrShared, self).has_permission(request, view) or
-                ('permission_link' in request.query_params and request.query_params['permission_link']))
+
+        permission_link = request.query_params.get('permission_link', None)
+        shipment_pk = request.parser_context['kwargs'].get('pk', None)
+
+        permission_obj = None
+        if permission_link:
+            try:
+                permission_obj = PermissionLink.objects.get(pk=permission_link)
+            except ObjectDoesNotExist:
+                pass
+        if (not permission_obj or not permission_obj.is_valid) and not shipment_pk:
+            return super(IsOwnerOrShared, self).has_permission(request, view)
+
+        if shipment_pk:
```

Comment:
would it make sense to maybe move this logic to before line 38? I think this is what's going to be hit most often

Reply:
You are right but the if clause just before won't be executed most of the time.